main.py

```diff
@@ -31,11 +38,194 @@ def setup(self):
+            # Changing to default pipeline. Miss flow entries are not needed
+            self.delete_miss_flows()
+        flows_by_swich = response.json()
+        set_up = self.build_content(pipeline)
+
+        for switch in flows_by_swich:
+            delete_flows = {"flows": []}
+            install_flows = {"flows": []}
+            for flow in flows_by_swich[switch]:
+                owner = flow["flow"]["owner"]
+                if owner not in set_up:
+                    continue
+                expected_table_id = set_up[owner][flow["flow"]["table_group"]]
+                # if table_id needs to change
+                if expected_table_id != flow["flow"]["table_id"]:
+                    # Get key-value from flow to be sent to flow_manager
+                    delete = {key: flow["flow"].get(key) for key in keys}
+                    delete["cookie_mask"] = flow["flow"]["cookie"]
+                    delete_flows["flows"].append(delete.copy())
+
+                    # Change table_id before being added
+                    flow["flow"].update({"table_id": expected_table_id})
+                    install_flows["flows"].append(flow["flow"].copy())
+
+            if delete_flows["flows"]:
+                self.send_flows(delete_flows, switch, "delete")
+            if install_flows["flows"]:
+                self.send_flows(install_flows, switch, "add")
+            self.install_miss_flows(pipeline, switch)
+        if pipeline.get('id'):
+            self.pipeline_controller.update_status(pipeline["id"], "enabled")
+
+    def install_miss_flows(self, pipeline: dict, switch: str):
+        """Install miss flow entry to a switch"""
+        install_flows = {"flows": []}
+        cookie = self.get_cookie(switch)
+        for table in pipeline["multi_table"][:-1]:
+            instruction = [{"instruction_type": "goto_table",
+                            "table_id": table['table_id'] + 1}]
+            flow = {
+                'priority': table.get('priority', 0),
+                'cookie': cookie,
+                'instructions': table.get('instructions', instruction),
+                'owner': 'of_multi_table',
+                'table_group': 'base',
+                'table_id': table['table_id'],
+            }
+            if table.get('match'):
+                flow['match'] = table.get('match')
+            install_flows["flows"].append(flow)
+
+        if install_flows["flows"]:
+            self.send_flows(install_flows, switch, "add")
+
+    def delete_miss_flows(self):
+        """Delete miss flows, aka. of_multi_table flows.
+        This method is called when returning to default pipeline."""
+        flows = {
+          "flows": [
+            {
+              "cookie": 12465963768561532928,
+              "cookie_mask": 18374686479671623680
+            }
+          ]
+        }
+        self.send_flows(flows, '', 'delete')
+
+    def send_flows(self, flows: Dict[str, list], switch: str, action: str):
```

Comment:
You can replace this with the event-oriented interface and remove `requests` entirely from the dependencies too, check out my previous comments and if you agree with it, all things considered the event-oriented interface can simplify for this NApp which will be dealing with a large number of flows.

Reply:
Changed to events